
# Laboratorio 3 ( Clasificación multiclase) Grupo 1
 <h3>Implementacion del modelo de regresion logistica para realizar predicciones con multiples clases.<h3>
 <HR>
 <h3>
  NOMBRE: Delgadillo LLanos Juan Sebastian <br>
  CARRERA: Ingenieria de Sistemas <BR>
  
  * [Enlace al git hub](https://github.com/sebastianDLL/SIS420_IA/tree/main)
  
  * [Enlace al Colab](https://colab.research.google.com/drive/1Z9DW73A1uVZbKNB9We8_ejAjqLGn10y-?usp=sharing)
 <h3>

In [86]:
# Nos conectamos a nuestro drive
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


## Vamos a utilizar el dataset Score.csv

### Temática:  Se quiere construir un modelo para clasificar a las personas en categorias segun sus datos crediticios

## Y --> Credit_Score : representa el rango de puntaje crediticio (pobre = 0, estándar = 1, bueno = 2)   

# ***PREPROCESAMIENTO DE NUESTRO DATASET***

In [87]:
# utilizado para la manipulación de directorios y rutas
import os

# Cálculo científico y vectorial para python
import numpy as np

# Libreria para graficos
from matplotlib import pyplot

# Modulo de optimizacion en scipy
from scipy import optimize

# modulo para cargar archivos en formato MATLAB
from scipy.io import loadmat

from sklearn.model_selection import train_test_split

# le dice a matplotlib que incruste gráficos en el cuaderno
%matplotlib inline

import pandas as pd

In [88]:
# La entrada es de 13 elementos (total)
input_layer_size  = 13
#clases o salidas de y + 1
num_labels = 4

#Haciendo uso de la libreria pandas para leer el dataset.
data = pd.read_csv('/content/drive/MyDrive/SIS420/LABORATORIOS/Lab3/Score.csv', delimiter=';')

#Imprimimos en una tabla el dataset.
data


Delay_from_due_date  Num_of_Delayed_Payment  Num_Credit_Inquiries  \
0                      3.0                     7.0                   4.0   
1                      3.0                     7.0                   4.0   
2                      3.0                     7.0                   4.0   
3                      5.0                     4.0                   4.0   
4                      6.0                     4.0                   4.0   
...                    ...                     ...                   ...   
25294                 19.0                    12.0                   8.0   
25295                 29.0                    13.0                   8.0   
25296                 21.0                    20.0                  11.0   
25297                 21.0                    21.0                  11.0   
25298                 22.0                    20.0                  11.0   

       Credit_History_Age   Age Annual_Income  Num_Bank_Accounts  \
0                   265.0  23.0      19114.12                3.0   
1                   265.0  23.0      19114.12                3.0   
2                   267.0  23.0      19114.12                3.0   
3                   268.0  23.0      19114.12                3.0   
4                   269.0  23.0      19114.12                3.0   
...                   ...   ...           ...                ...   
25294               111.0  44.0      70350.06                5.0   
25295               111.0  44.0      70350.06                5.0   
25296               115.0  31.0      48597.42                9.0   
25297               116.0  31.0      48597.42                9.0   
25298               116.0  31.0      48597.42                9.0   

       Num_Credit_Card  Interest_Rate  Num_of_Loan  Changed_Credit_Limit  \
0                  4.0            3.0          4.0                 11.27   
1                  4.0            3.0          4.0                 11.27   
2                  4.0            3.0          4.0                 11.27   
3                  4.0            3.0          4.0                 11.27   
4                  4.0            3.0          4.0                 11.27   
...                ...            ...          ...                   ...   
25294              3.0           12.0          2.0                  8.23   
25295              3.0           12.0          2.0                  8.23   
25296              8.0           30.0          3.0                  2.61   
25297              8.0           30.0          3.0                  2.61   
25298              8.0           30.0          3.0                  2.61   

       Outstanding_Debt Credit_Score  
0                809.98         Good  
1                809.98         Good  
2                809.98         Good  
3                809.98         Good  
4                809.98         Good  
...                 ...          ...  
25294            148.52     Standard  
25295            148.52     Standard  
25296           1868.46         Poor  
25297           1868.46         Poor  
25298           1868.46         Poor  

[25299 rows x 13 columns]

In [89]:
#CONVERTIMOS NUESTRAS (Y) A VALORES NUMERICOS

# Definir un diccionario de mapeo de etiquetas literales a enteros
label_mapping = {'Good': 2, 'Standard': 1, 'Poor': 0}

# Mapear las etiquetas literales a enteros
data['Credit_Score'] = data['Credit_Score'].map(label_mapping)

# Verificar que se haya realizado correctamente
data


Delay_from_due_date  Num_of_Delayed_Payment  Num_Credit_Inquiries  \
0                      3.0                     7.0                   4.0   
1                      3.0                     7.0                   4.0   
2                      3.0                     7.0                   4.0   
3                      5.0                     4.0                   4.0   
4                      6.0                     4.0                   4.0   
...                    ...                     ...                   ...   
25294                 19.0                    12.0                   8.0   
25295                 29.0                    13.0                   8.0   
25296                 21.0                    20.0                  11.0   
25297                 21.0                    21.0                  11.0   
25298                 22.0                    20.0                  11.0   

       Credit_History_Age   Age Annual_Income  Num_Bank_Accounts  \
0                   265.0  23.0      19114.12                3.0   
1                   265.0  23.0      19114.12                3.0   
2                   267.0  23.0      19114.12                3.0   
3                   268.0  23.0      19114.12                3.0   
4                   269.0  23.0      19114.12                3.0   
...                   ...   ...           ...                ...   
25294               111.0  44.0      70350.06                5.0   
25295               111.0  44.0      70350.06                5.0   
25296               115.0  31.0      48597.42                9.0   
25297               116.0  31.0      48597.42                9.0   
25298               116.0  31.0      48597.42                9.0   

       Num_Credit_Card  Interest_Rate  Num_of_Loan  Changed_Credit_Limit  \
0                  4.0            3.0          4.0                 11.27   
1                  4.0            3.0          4.0                 11.27   
2                  4.0            3.0          4.0                 11.27   
3                  4.0            3.0          4.0                 11.27   
4                  4.0            3.0          4.0                 11.27   
...                ...            ...          ...                   ...   
25294              3.0           12.0          2.0                  8.23   
25295              3.0           12.0          2.0                  8.23   
25296              8.0           30.0          3.0                  2.61   
25297              8.0           30.0          3.0                  2.61   
25298              8.0           30.0          3.0                  2.61   

       Outstanding_Debt  Credit_Score  
0                809.98             2  
1                809.98             2  
2                809.98             2  
3                809.98             2  
4                809.98             2  
...                 ...           ...  
25294            148.52             1  
25295            148.52             1  
25296           1868.46             0  
25297           1868.46             0  
25298           1868.46             0  

[25299 rows x 13 columns]

### SEPARAMOS EL 80 % PARA ENTRENAR Y EL 20% PARA PROBAR


In [90]:
# Separar el conjunto de datos en conjuntos de entrenamiento y prueba
train_dataset, test_dataset = train_test_split(data, test_size=0.2, random_state=42)

# Separar los conjuntos de características (X) y las etiquetas (y) para el conjunto de prueba
X_test = test_dataset.drop(['Credit_Score'], axis=1)
y_test = test_dataset['Credit_Score']

# Separar los conjuntos de características (X) y las etiquetas (y) para el conjunto de entrenamiento
X_train = train_dataset.drop(['Credit_Score'], axis=1)
y_train = train_dataset['Credit_Score']

# Calcular el número de ejemplos de entrenamiento
m_train = len(y_train)

print(X_train)
print(y_train)

# Imprimir la cantidad de ejemplos de entrenamiento y prueba
print(" ")
print('El 80% de ejemplos para entrenamiento son la cantidad de: {:.0f} de ejemplos'.format(len(train_dataset)))
print('El 20% de ejemplos para pruebas son la cantidad de: {:.0f} de ejemplos'.format(len(test_dataset)))



       Delay_from_due_date  Num_of_Delayed_Payment  Num_Credit_Inquiries  \
2458                  15.0                    21.0                   9.0   
23941                 10.0                     9.0                   7.0   
22437                 12.0                    18.0                   8.0   
19910                  5.0                     7.0                   8.0   
7302                   4.0                     5.0                   0.0   
...                    ...                     ...                   ...   
21575                 26.0                    16.0                   7.0   
5390                  24.0                     9.0                   2.0   
860                   12.0                     9.0                   2.0   
15795                 48.0                    22.0                   8.0   
23654                 11.0                    20.0                   1.0   

       Credit_History_Age   Age        Annual_Income  Num_Bank_Accounts  \
2458        

In [91]:
# Imprimir las características (X) de un ejemplo específico
print("Características (X) del primer ejemplo:")
print(X_train.iloc[0, :])  # Esto imprimirá todas las características del primer ejemplo en X_train

# Imprimir la etiqueta (y) correspondiente al primer ejemplo
print("Etiqueta (y) del primer ejemplo:")
print(y_train.iloc[0])  # Esto imprimirá la etiqueta del primer ejemplo en y_train


Características (X) del primer ejemplo:
Delay_from_due_date           15.0
Num_of_Delayed_Payment        21.0
Num_Credit_Inquiries           9.0
Credit_History_Age           155.0
Age                           42.0
Annual_Income             69692.16
Num_Bank_Accounts              9.0
Num_Credit_Card                5.0
Interest_Rate                 32.0
Num_of_Loan                    7.0
Changed_Credit_Limit         14.66
Outstanding_Debt           2128.87
Name: 2458, dtype: object
Etiqueta (y) del primer ejemplo:
1


In [92]:
def  featureNormalize(X):
    X_norm = X.copy()
    mu = np.zeros(X.shape[1])
    sigma = np.zeros(X.shape[1])

    mu = np.mean(X, axis = 0)
    sigma = np.std(X, axis = 0)
    X_norm = (X - mu) / sigma

    return X_norm, mu, sigma

Faltaban Datos asi que rellenamos con la media de la columna {Annual_Income}
Luego Normalizamos

In [93]:
# Limpiar los valores de la columna Annual_Income
X_train['Annual_Income'] = X_train['Annual_Income'].str.replace(',', '').astype(float)

# Calcular la media de la columna Annual_Income
mean_annual_income = X_train['Annual_Income'].mean()

# Imputar los valores faltantes con la media
X_train['Annual_Income'].fillna(mean_annual_income, inplace=True)

# Llama featureNormalize con los datos cargados
X_norm, mu, sigma = featureNormalize(X_train)


In [94]:
X_norm

Delay_from_due_date  Num_of_Delayed_Payment  Num_Credit_Inquiries  \
2458             -0.420448                1.221819              0.821585   
23941            -0.763751               -0.694343              0.310864   
22437            -0.626430                0.742778              0.566225   
19910            -1.107053               -1.013704              0.566225   
7302             -1.175714               -1.333064             -1.476661   
...                    ...                     ...                   ...   
21575             0.334818                0.423418              0.310864   
5390              0.197497               -0.694343             -0.965939   
860              -0.626430               -0.694343             -0.965939   
15795             1.845350                1.381499              0.566225   
23654            -0.695090                1.062139             -1.221300   

       Credit_History_Age       Age  Annual_Income  Num_Bank_Accounts  \
2458            -0.674128  0.811820      -0.132253           1.396754   
23941            1.347846  0.344148      -0.132253          -0.901687   
22437           -1.220060 -1.713607      -0.132253           0.247534   
19910            0.832242  1.092423      -0.132253          -0.901687   
7302             1.630922 -1.339469      -0.132253          -0.518613   
...                   ...       ...            ...                ...   
21575           -0.805556 -0.684729      -0.132253           1.779827   
5390             0.417738 -0.778263       5.851977           0.247534   
860             -0.219184  2.027766      -0.132253          -1.667833   
15795           -0.734787 -0.778263      -0.132253           1.396754   
23654            1.084989  1.747163      -0.132253           1.013681   

       Num_Credit_Card  Interest_Rate  Num_of_Loan  Changed_Credit_Limit  \
2458         -0.257875       1.998209     1.411985              0.670658   
23941        -1.233778      -1.406820    -1.463394             -1.297066   
22437         2.181882       0.182194    -0.231089              0.450310   
19910        -0.745826      -1.293319     0.179680             -0.469605   
7302          0.230077      -0.952816    -1.052626             -1.041278   
...                ...            ...          ...                   ...   
21575         1.205979       0.749699     1.001216              0.826288   
5390         -1.233778      -1.179818    -1.052626             -0.170671   
860           0.718028      -1.293319    -0.231089             -0.987346   
15795         0.718028       2.111710     1.001216             -0.558977   
23654         0.718028      -0.158309    -1.052626              0.769275   

       Outstanding_Debt  
2458           0.617095  
23941         -0.499166  
22437          0.306180  
19910         -0.377016  
7302          -0.323707  
...                 ...  
21575          0.967457  
5390          -0.306882  
860           -0.915354  
15795          0.083569  
23654         -0.734595  

[20239 rows x 12 columns]

In [95]:
# Obtener las dimensiones de la matriz de características X_train
m, n = X_train.shape

# X_train ya ha sido normalizado previamente, por lo que simplemente asignamos la versión normalizada a X_train
X_train = X_norm

# Las dimensiones de la matriz de características X_train se utilizan en el proceso de entrenamiento del modelo
# m representa el número de ejemplos de entrenamiento
# n representa el número de características
# Al asignar X_norm a X_train, nos aseguramos de que los datos de entrenamiento estén normalizados y listos para su uso en el modelo


In [96]:
# Seleccionar aleatoriamente 100 puntos de datos para mostrar
# Se utiliza np.random.choice para seleccionar aleatoriamente 100 índices de entre 0 y m-1, sin reemplazo
# donde m es el número total de ejemplos de entrenamiento
rand_indices = np.random.choice(m, 100, replace=False)

# Seleccionar las filas correspondientes a los índices aleatorios de X_train
# Utilizamos iloc para acceder a las filas utilizando los índices seleccionados aleatoriamente
sel = X_train.iloc[rand_indices, :]

# La variable 'sel' ahora contiene 100 ejemplos de datos de entrenamiento seleccionados aleatoriamente


# FUNCION SIGMOIDEAL

In [97]:
def sigmoid(z):
    # Calcula la sigmoide de una entrada z
    # convierte la intrada a un arreglo numpy
    z = np.array(z)

    g = np.zeros(z.shape)

    g = 1 / (1 + np.exp(-z))

    return g


### Esta función lrCostFunction calcula el costo y el gradiente para la regresión logística regularizada

In [98]:
def lrCostFunction(theta, X, y, lambda_):
    """
    Calcula el costo de usar theta como parámetro para la regresión logística regularizada y
    el gradiente del costo w.r.t. a los parámetros.

    Parametros
    ----------
    theta : array_like
        Parametro theta de la regresion logistica. Vector de la forma(shape) (n, ). n es el numero de caracteristicas
        incluida la intercepcion

    X : array_like
        Dataset con la forma(shape) (m x n). m es el numero de ejemplos, y n es el numero de
        caracteristicas (incluida la intercepcion).

    y : array_like
        El conjunto de etiquetas. Un vector con la forma (shape) (m, ). m es el numero de ejemplos

    lambda_ : float
        Parametro de regularización.

    Devuelve
    -------
    J : float
        El valor calculado para la funcion de costo regularizada.

    grad : array_like
        Un vector de la forma (shape) (n, ) que es el gradiente de la
        función de costo con respecto a theta, en los valores actuales de theta..
    """

    # Inicializa algunos valores utiles
    m = y.size

    # convierte las etiquetas a valores enteros si son boleanos
    if y.dtype == bool:
        y = y.astype(int)

    J = 0
    grad = np.zeros(theta.shape)

    h = sigmoid(X.dot(theta.T))

    temp = theta
    temp[0] = 0

#     J = (1 / m) * np.sum(-y.dot(np.log(h)) - (1 - y).dot(np.log(1 - h)))
    J = (1 / m) * np.sum(-y.dot(np.log(h)) - (1 - y).dot(np.log(1 - h))) + (lambda_ / (2 * m)) * np.sum(np.square(temp))

    grad = (1 / m) * (h - y).dot(X)
#     theta = theta - (alpha / m) * (h - y).dot(X)
    grad = grad + (lambda_ / m) * temp

    return J, grad
#    return J, theta

### Entrena clasificadores de regresión logística para cada clase y devuelve los parámetros entrenados para cada uno de estos clasificadores en una matriz all_theta.

In [99]:
def oneVsAll(X, y, num_labels, lambda_):
    """
    Trains num_labels logistic regression classifiers and returns
    each of these classifiers in a matrix all_theta, where the i-th
    row of all_theta corresponds to the classifier for label i.

    Parameters
    ----------
    X : array_like
        The input dataset of shape (m x n). m is the number of
        data points, and n is the number of features. Note that we
        do not assume that the intercept term (or bias) is in X, however
        we provide the code below to add the bias term to X.

    y : array_like
        The data labels. A vector of shape (m, ).

    num_labels : int
        Number of possible labels.

    lambda_ : float
        The logistic regularization parameter.

    Returns
    -------
    all_theta : array_like
        The trained parameters for logistic regression for each class.
        This is a matrix of shape (K x n+1) where K is number of classes
        (ie. `numlabels`) and n is number of features without the bias.
    """
    # algunas variables utiles
    m, n = X.shape

    all_theta = np.zeros((num_labels, n + 1))

    # Agrega unos a la matriz X
    X = np.concatenate([np.ones((m, 1)), X], axis=1)

    for c in np.arange(num_labels):
        initial_theta = np.zeros(n + 1)
        options = {'maxiter': 50}
        res = optimize.minimize(lrCostFunction,
                                initial_theta,
                                (X, (y == c), lambda_),
                                jac=True,
                                method='CG',
                                options=options)

        all_theta[c] = res.x

    return all_theta

In [100]:

print(X_norm)


       Delay_from_due_date  Num_of_Delayed_Payment  Num_Credit_Inquiries  \
2458             -0.420448                1.221819              0.821585   
23941            -0.763751               -0.694343              0.310864   
22437            -0.626430                0.742778              0.566225   
19910            -1.107053               -1.013704              0.566225   
7302             -1.175714               -1.333064             -1.476661   
...                    ...                     ...                   ...   
21575             0.334818                0.423418              0.310864   
5390              0.197497               -0.694343             -0.965939   
860              -0.626430               -0.694343             -0.965939   
15795             1.845350                1.381499              0.566225   
23654            -0.695090                1.062139             -1.221300   

       Credit_History_Age       Age  Annual_Income  Num_Bank_Accounts  \
2458          

In [101]:
# Definir el parámetro de regularización lambda
lambda_ = 0.1

# Aplicar el enfoque de uno contra todos para entrenar el clasificador
# La función oneVsAll entrenará un clasificador para cada clase en el conjunto de datos
# y devolverá los parámetros theta (pesos) correspondientes a cada clase
all_theta = oneVsAll(X_train, y_train, num_labels, lambda_)

# Imprimir la forma de la matriz de parámetros theta
print(all_theta.shape)


(4, 13)


### Mostramos nuestras THETAS

In [102]:
print(all_theta)

[[-1.12093218e+00  2.38024427e-01 -1.69627369e-01  3.42194669e-01
  -1.83583772e-01 -1.47982381e-02  3.62474578e-02  2.94468220e-02
   3.63677205e-01  4.40430461e-01  3.20220479e-02 -5.17025897e-01
   2.20034635e-01]
 [ 1.35468584e-01 -1.09853656e-01  3.17198759e-01 -2.24956228e-01
  -1.53511177e-02 -3.39873170e-02  1.67090570e-02  1.31523588e-01
  -2.80682081e-02 -1.19289316e-01 -7.77810153e-02  5.77877776e-01
  -4.54815233e-01]
 [-2.57053973e+00 -5.54687266e-01 -1.66984054e-01 -1.92916672e-01
   2.57898930e-01  8.15525771e-02 -5.70919505e-02 -1.41299399e-01
  -5.20062621e-01 -7.17751290e-01  3.55369552e-05 -2.66113240e-01
   3.18175468e-02]
 [-1.24380805e+01 -7.90596957e-11 -7.68096900e-11 -8.39631347e-11
   8.17211562e-11  3.50576188e-11  1.83972738e-12 -7.79647411e-11
  -6.77350077e-11 -8.61038528e-11 -8.03760505e-11 -5.86412698e-11
  -8.60900303e-11]]


### Realiza predicciones para cada ejemplo en la matriz X utilizando los parámetros entrenados all_theta para todos los clasificadores de regresión logística.

In [103]:
def predictOneVsAll(all_theta, X):
    """
    Devuelve un vector de predicciones para cada ejemplo en la matriz X.
    Tenga en cuenta que X contiene los ejemplos en filas.
    all_theta es una matriz donde la i-ésima fila es un vector theta de regresión logística entrenada para la i-ésima clase.
    Debe establecer p en un vector de valores de 0..K-1 (por ejemplo, p = [0, 2, 0, 1]
    predice clases 0, 2, 0, 1 para 4 ejemplos).

    Parametros
    ----------
    all_theta : array_like
        The trained parameters for logistic regression for each class.
        This is a matrix of shape (K x n+1) where K is number of classes
        and n is number of features without the bias.

    X : array_like
        Data points to predict their labels. This is a matrix of shape
        (m x n) where m is number of data points to predict, and n is number
        of features without the bias term. Note we add the bias term for X in
        this function.

    Devuelve
    -------
    p : array_like
        The predictions for each data point in X. This is a vector of shape (m, ).
    """

    m = X.shape[0];
    num_labels = all_theta.shape[0]

    p = np.zeros(m)

    # Add ones to the X data matrix
    X = np.concatenate([np.ones((m, 1)), X], axis=1)
    p = np.argmax(sigmoid(X.dot(all_theta.T)), axis = 1)

    return p

In [104]:
# Imprime la forma de la matriz de características X_train
print(X_train.shape)

# Utiliza los parámetros all_theta para predecir las etiquetas de clase para el conjunto de entrenamiento X_train
pred = predictOneVsAll(all_theta, X_train)

# Calcula la precisión del conjunto de entrenamiento comparando las predicciones con las etiquetas verdaderas
print('Precision del conjuto de entrenamiento: {:.2f}%'.format(np.mean(pred == y_train) * 100))

# Crea una copia del subconjunto de características X_train llamado XPrueba seleccionando las filas de la 10 a la 150
XPrueba = X_train.iloc[10:150, :].copy()

# Imprime la forma de la matriz XPrueba
print(XPrueba.shape)

# Añade una columna de unos a la matriz XPrueba para considerar el término de sesgo
XPrueba = np.concatenate([np.ones((140, 1)), XPrueba], axis=1)

# Utiliza los parámetros all_theta para predecir las etiquetas de clase para el conjunto XPrueba
p = np.argmax(sigmoid(XPrueba.dot(all_theta.T)), axis=1)
print(p)

# Imprime las etiquetas verdaderas del conjunto de entrenamiento, seleccionando las filas de la 10 a la 150
print(y_train.iloc[10:150])



(20239, 12)
Precision del conjuto de entrenamiento: 63.04%
(140, 12)
[1 1 1 1 0 1 1 1 1 0 1 2 0 1 1 0 0 1 0 1 2 1 1 1 2 2 1 1 2 1 1 1 1 0 1 1 1
 1 1 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 0 0 1 2 2 1 1 1 1 1 1 2 1 1 2 0 1 0
 1 0 1 1 1 0 1 2 0 1 0 1 1 0 1 1 0 1 0 1 1 1 1 0 1 1 1 1 1 1 1 0 1 0 1 1 1
 2 0 1 0 0 1 1 1 2 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 0]
24822    1
5199     1
11522    2
8375     2
6224     0
        ..
8403     1
13681    1
18272    1
18181    0
10434    1
Name: Credit_Score, Length: 140, dtype: int64


## VALIDACION CON NUESTRO CONNJUNTO DE PRUEBA

In [105]:
print("Forma de X:", X_test.shape)
print("Forma de all_theta:", all_theta.shape)


Forma de X: (5060, 12)
Forma de all_theta: (4, 13)


In [106]:
#vemos informacion de el conjunto de X para pruebas
print(X_test.info())


<class 'pandas.core.frame.DataFrame'>
Int64Index: 5060 entries, 25131 to 5405
Data columns (total 12 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Delay_from_due_date     5060 non-null   float64
 1   Num_of_Delayed_Payment  5060 non-null   float64
 2   Num_Credit_Inquiries    5060 non-null   float64
 3   Credit_History_Age      5060 non-null   float64
 4   Age                     5060 non-null   float64
 5   Annual_Income           5060 non-null   object 
 6   Num_Bank_Accounts       5060 non-null   float64
 7   Num_Credit_Card         5060 non-null   float64
 8   Interest_Rate           5060 non-null   float64
 9   Num_of_Loan             5060 non-null   float64
 10  Changed_Credit_Limit    5060 non-null   float64
 11  Outstanding_Debt        5060 non-null   float64
dtypes: float64(11), object(1)
memory usage: 513.9+ KB
None


Podemos observar que es de tipo objeto convertiremos a float  (5   Annual_Income           5060 non-null   object)

In [107]:
X_test['Annual_Income'] = pd.to_numeric(X_test['Annual_Income'], errors='coerce')


Normalizamos el conjunto de pruebas

In [108]:
#FALTA NORMALIZAR
X_norm_test = (X_test- mu) / sigma

m_test= len(X_test)
X_norm_test

Delay_from_due_date  Num_of_Delayed_Payment  Num_Credit_Inquiries  \
25131            -1.038393                0.104058             -0.965939   
1246             -0.626430                0.742778              0.310864   
7964             -1.244375               -1.652424             -0.710579   
1610             -0.695090               -0.534663             -0.965939   
10491            -1.107053                0.263738              0.055503   
...                    ...                     ...                   ...   
20794            -0.008485               -0.055623             -0.455218   
8623             -0.283127               -0.055623             -0.965939   
10727             0.746781                0.423418              0.821585   
1163             -1.175714               -1.492744             -1.476661   
5405              0.197497                1.062139              1.076946   

       Credit_History_Age       Age  Annual_Income  Num_Bank_Accounts  \
25131            1.539933 -1.245935      -0.132253          -0.518613   
1246             0.023453 -0.029989      -0.132253           1.013681   
7964             0.599716  0.811820      -0.132253          -1.667833   
1610             0.397518 -0.310592      -0.132253           0.247534   
10491           -1.533466 -0.965332            NaN          -0.135540   
...                   ...       ...            ...                ...   
20794           -0.451711  0.344148      -0.132253           1.013681   
8623             1.014220 -0.591195      -0.132253           1.013681   
10727           -2.008630 -0.029989      -0.132253           0.630607   
1163            -0.289953  0.905354      -0.132253          -2.050907   
5405            -0.845995  0.718286      -0.132253           1.396754   

       Num_Credit_Card  Interest_Rate  Num_of_Loan  Changed_Credit_Limit  \
25131        -0.257875      -0.385311     0.179680              1.145254   
1246         -1.233778       0.522697    -0.231089              0.744621   
7964         -1.233778      -0.271810    -0.231089             -1.203072   
1610         -1.233778      -0.612313     0.179680              0.527354   
10491         0.718028       0.295695    -1.052626              0.051218   
...                ...            ...          ...                   ...   
20794        -1.233778      -0.385311    -0.641857              1.253116   
8623          0.718028      -1.066317     0.179680             -1.244676   
10727         0.230077       1.544205     2.233522              2.590614   
1163         -2.209680      -0.385311    -0.231089             -0.708444   
5405          1.693931       1.544205     2.233522              2.643004   

       Outstanding_Debt  
25131         -1.084863  
1246          -0.206942  
7964          -0.723413  
1610          -1.162711  
10491         -0.022167  
...                 ...  
20794          1.059451  
8623          -0.289877  
10727          2.936962  
1163          -0.029052  
5405           1.477283  

[5060 rows x 12 columns]

In [109]:
y_test

25131    1
1246     1
7964     1
1610     1
10491    1
        ..
20794    1
8623     1
10727    1
1163     2
5405     0
Name: Credit_Score, Length: 5060, dtype: int64

### Realizamos predicciones con nuestro conjunto de datos de prueba y evalualuamos la precisión del modelo entrenado.

In [110]:
print(X_test.shape)
# Llama a la función predictOneVsAll con los datos de prueba
pred = predictOneVsAll(all_theta, X_norm_test)
print('Precision del conjuto de entrenamiento: {:.2f}%'.format(np.mean(pred == y_test) * 100))
XPrueba = X_norm_test.iloc[10:450, :].copy()
print(XPrueba.shape)
#print(np.ones((1)))
#print(XPrueba)
#p = np.zeros(1)
XPrueba = np.concatenate([np.ones((440, 1)), XPrueba], axis=1)
print(XPrueba.shape)
p = np.argmax(sigmoid(XPrueba.dot(all_theta.T)), axis = 1)
print(p)

# displayData(X[1002:1003, :])
y_test[10:450]

(5060, 12)
Precision del conjuto de entrenamiento: 58.54%
(440, 12)
(440, 13)
[2 1 1 0 0 1 1 2 0 1 1 2 0 0 0 0 2 1 0 0 1 1 1 1 2 0 1 1 1 1 0 1 1 0 2 1 2
 1 1 1 0 1 0 1 2 0 2 1 1 1 0 1 0 1 0 1 0 2 1 1 0 1 0 1 1 0 1 1 1 1 1 1 0 1
 1 1 1 2 0 2 0 1 1 1 0 2 1 0 0 2 1 2 1 1 1 0 0 0 1 1 0 0 1 0 1 1 2 1 0 1 2
 0 1 0 1 1 2 1 0 2 1 0 0 1 0 1 1 0 2 0 1 0 0 0 1 1 1 0 1 0 0 1 1 0 0 1 0 0
 1 1 1 0 0 0 1 0 0 2 1 1 0 1 1 0 1 1 1 0 1 1 1 0 1 0 1 0 1 1 1 1 0 2 0 1 1
 1 1 0 1 1 1 0 0 0 1 2 0 1 1 1 0 1 0 0 0 1 0 1 1 0 1 1 1 1 1 0 1 1 1 1 1 1
 0 0 0 0 0 0 0 2 1 2 1 1 0 2 1 1 1 1 1 1 0 1 2 1 1 1 1 0 1 1 1 1 1 0 1 1 2
 1 0 1 1 2 1 1 1 0 0 2 0 1 2 1 1 1 1 1 0 1 0 2 1 1 1 1 0 0 1 1 1 0 1 1 1 1
 0 1 0 1 2 0 1 0 0 1 2 2 1 0 0 2 1 1 1 1 1 1 0 0 0 1 1 0 0 2 1 1 1 1 1 0 0
 1 0 2 1 1 1 2 1 0 1 1 1 2 1 1 1 0 0 0 1 1 0 1 0 0 0 1 1 0 1 1 1 1 1 1 2 2
 1 1 0 2 1 0 1 1 1 1 1 0 0 1 1 1 1 1 0 1 1 1 1 1 0 1 0 1 1 1 1 1 1 0 1 1 0
 1 0 1 1 2 0 0 1 1 0 1 1 1 1 1 1 1 0 1 1 1 0 0 1 1 1 1 1 1 1 1 0 1]


6275     1
11838    0
7481     1
5325     1
4834     1
        ..
23844    1
5050     2
13745    2
10119    2
19165    1
Name: Credit_Score, Length: 440, dtype: int64